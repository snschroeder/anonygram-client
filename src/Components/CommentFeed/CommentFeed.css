.CommentFeed__form {
  width: 100%;
  height: 3em;

}

.CommentFeed__input {
  width: 80%;
  height: 100%;
  padding: 1em;
  box-shadow: rgba(0, 0, 0, 0.15) -6px 6px 6px -6px, rgba(0, 0, 0, 0.1) 0px 0px 1px -1px;
  border-radius: 5px;
}

.CommentFeed__button {
  width: 20%;
  height: 100%;
  background-color: #80c683;
  box-shadow: rgba(0, 0, 0, 0.15) -6px 6px 6px -6px, rgba(0, 0, 0, 0.1) 0px 0px 1px -1px;
  padding: 1em;
  border-radius: 5px;
}


