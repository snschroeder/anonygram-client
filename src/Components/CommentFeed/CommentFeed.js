import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import randomizer from '../../helpers/randomizer';
import Comment from './Comment/Comment';
import CommentApi from '../../services/comment-api-service';
import TokenService from '../../services/token-service';
import UserContext from '../../contexts/UserContext';
import './CommentFeed.css';

import moment from 'moment';

export default class CommentFeed extends Component {
  state = {
    usernames: {},
    newComment: '',
  }
  
  static contextType = UserContext;

  static defaultProps = {
    comments: []
  }

  processTimestamp(timestamp) {
    return moment(new Date(timestamp)).local().fromNow();
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const commentText = this.state.newComment; 
    const { id } = TokenService.parseAuthToken()
    CommentApi.postComment(this.props.id, commentText, id)
      .then(res => {
        this.props.setCommentsByPush(res);
        //If first post by this user, a username will need to be generated for them
        if (!this.state.usernames[id]) {
          const newUn = this.generateUsernames([res]);
          const newUsernames = {...this.state.usernames, ...newUn}
          this.setState({ usernames: newUsernames, newComment: '' })
        } else {
          this.setState({ newComment: '' })
        }
      })
  }

  handleChange = (e) => {
    this.setState({newComment: e.target.value})
  }

  componentDidMount() {
    if (this.props.comments.length > 0) {
      const newUsernames = this.generateUsernames(this.props.comments)
      this.setState({ usernames: newUsernames });
    }
  };

  generateUsernames = (comments) => {
    return randomizer.getAnonUsernames(comments);
  }

  render() {
    const { comments, userLoggedIn } = this.props;

    //If you're posting the first comment, then generateUsernames didn't run on mount. Run it here. 
    //ComponentDidMount edited to be >1 so that generateUsernames doesn't run twice if there is one 
    //comment to begin with
    // (this.props.comments.length === 1 && (Object.keys(this.state.usernames).length === 0)) && this.generateUsernames(this.props.comments);

    return (
      <div className='comment-container'>
        {(comments.length === 0) 
          ? null
          : (
            comments.map(commentObj => {
              const { comment_timestamp, user_id, comment_id, comment_text } = commentObj;
    
              return <Comment key={comment_id} text={comment_text} username={this.state.usernames[user_id]} timestamp={this.processTimestamp(comment_timestamp)} />;
            })
          )
        }
        <form onSubmit={(e) => this.handleSubmit(e)} className='CommentFeed__form'>
          {/* <label htmlFor='newComment'>Add a Comment</label> */}
          { //Conditionally render for logged in user
            userLoggedIn ? (
              <>
                <input className='CommentFeed__input' onChange={(e) => this.handleChange(e)} value={this.state.newComment} type='text' id='newComment' placeholder='Add a comment...' />
                <button type='submit' className='CommentFeed__button' disabled={!this.state.newComment}>Post</button>
              </>
            ) : (
              <>
                <label htmlFor='newComment'>You must be logged in to post a comment. Log in <Link to='/login'>here</Link>.</label>
                <input type='text' id='newComment' className='CommentFeed__input' placeholder='' disabled />
                <button type='submit' className='CommentFeed__button' disabled>Post</button>
              </>
            )
          }
        </form>
      </div>
    )
  }
}

