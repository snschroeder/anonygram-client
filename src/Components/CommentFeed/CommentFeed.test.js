import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import CommentFeed from './CommentFeed';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('CommendFeed component', () => {
  CommentFeed.contextTypes = {
    user: PropTypes.object,
  };
  const context = {
    user: {
      id: 'b709597c-bfbd-498c-9daa-da18556afcb9',
    },
  };
  const props = {
    id: '4d8a8589-1c1d-48e8-97fa-f2cee900e1cf',
    comments: [
      {
        comment_timestamp: new Date(),
        user_id: '1',
        comment_id: '56532f9e-8022-4759-b134-b7269d510689',
        comment_text: 'comment text 1',
      },
    ],
    setCommentsByPush: () => {},
  };
  
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <CommentFeed />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders elements by default', () => {
    const wrapper = shallow(<CommentFeed />, { context });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders elements given props', () => {
    const wrapper = shallow(<CommentFeed {...props} />, { context });
    wrapper.setState({
      usernames: {
        '1': 'ObtuseRubberGoose'
      } 
    })
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('obscures usernames when multiple comments present', () => {
    const propsWithComments = {
      id: '4d8a8589-1c1d-48e8-97fa-f2cee900e1cf',
      comments: [
        {
          comment_timestamp: new Date(),
          user_id: '1',
          comment_id: '56532f9e-8022-4759-b134-b7269d510689',
          comment_text: 'comment text 1',
        },
        {
          comment_timestamp: new Date(),
          user_id: '2',
          comment_id: '56532f9e-8022-4759-b134-b7269d510689',
          comment_text: 'comment text 1',
        },
      ],
      setCommentsByPush: () => {},
    };
    const wrapper = shallow(<CommentFeed {...propsWithComments} />, {
      context,
    });
    const user1 = wrapper.childAt(0).props().username;
    const user2 = wrapper.childAt(1).props().username;
    expect(user1 === user2).toEqual(false);
  });
});
