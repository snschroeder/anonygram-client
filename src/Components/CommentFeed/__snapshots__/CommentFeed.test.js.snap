// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommendFeed component renders elements by default 1`] = `
<div
  className="comment-container"
>
  <form
    className="CommentFeed__form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="newComment"
    >
      You must be logged in to post a comment. Log in 
      <Link
        to="/login"
      >
        here
      </Link>
      .
    </label>
    <input
      className="CommentFeed__input"
      disabled={true}
      id="newComment"
      placeholder=""
      type="text"
    />
    <button
      className="CommentFeed__button"
      disabled={true}
      type="submit"
    >
      Post
    </button>
  </form>
</div>
`;

exports[`CommendFeed component renders elements given props 1`] = `
<div
  className="comment-container"
>
  <Comment
    key="56532f9e-8022-4759-b134-b7269d510689"
    text="comment text 1"
    timestamp="a few seconds ago"
    username="ObtuseRubberGoose"
  />
  <form
    className="CommentFeed__form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="newComment"
    >
      You must be logged in to post a comment. Log in 
      <Link
        to="/login"
      >
        here
      </Link>
      .
    </label>
    <input
      className="CommentFeed__input"
      disabled={true}
      id="newComment"
      placeholder=""
      type="text"
    />
    <button
      className="CommentFeed__button"
      disabled={true}
      type="submit"
    >
      Post
    </button>
  </form>
</div>
`;
